File: basic_samples/nodejs/score_task.js
/**
 * Speechace API — Task Achievement Scoring
 * =========================================
 * Submits an audio file and an image description for task achievement scoring.
 * Returns a task score (0-5) plus pronunciation, fluency, grammar, vocabulary,
 * and coherence metrics across multiple scoring frameworks.
 *
 * Usage:
 *   node score_task.js --audio ../../audio_samples/apple.wav \
 *       --context "A photo of a red apple on a white table"
 *
 * Environment variables (can also be placed in a .env file at the project root):
 *   SPEECHACE_API_KEY       Your Speechace API key (required)
 *   SPEECHACE_API_ENDPOINT  Base URL, defaults to https://api.speechace.co
 *   SPEECHACE_USER_ID       Arbitrary user identifier sent to the API
 */

import { createReadStream } from "fs";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";
import { parseArgs } from "util";
import FormData from "form-data";
import fetch from "node-fetch";
import dotenv from "dotenv";

const __dirname = dirname(fileURLToPath(import.meta.url));
dotenv.config({ path: resolve(__dirname, "../../.env") });

const ROWS = ["overall", "pronunciation", "fluency", "grammar", "vocab", "coherence"];

export function buildUrl(endpoint, key, dialect, userId) {
  const base = endpoint.replace(/\/$/, "");
  const params = new URLSearchParams({ key, dialect, user_id: userId, task_type: "describe-image" });
  return `${base}/api/scoring/task/v9/json?${params}`;
}

export async function scoreTask(audioPath, context, dialect = "en-us") {
  const key = process.env.SPEECHACE_API_KEY ?? "";
  const endpoint = process.env.SPEECHACE_API_ENDPOINT ?? "https://api.speechace.co";
  const userId = process.env.SPEECHACE_USER_ID ?? "sample-user-001";

  if (!key) {
    throw new Error(
      "SPEECHACE_API_KEY is not set. " +
      "Export it as an environment variable or add it to a .env file."
    );
  }

  const url = buildUrl(endpoint, key, dialect, userId);
  const form = new FormData();
  form.append("task_context", context);
  form.append("include_speech_score", "1");
  form.append("user_audio_file", createReadStream(audioPath));

  const response = await fetch(url, { method: "POST", body: form });
  if (!response.ok) {
    const body = await response.text();
    throw new Error(`HTTP ${response.status}: ${body}`);
  }
  return response.json();
}

export function printSummary(result) {
  if (result.status !== "success") {
    console.log(`API returned status: ${result.status}`);
    console.log(JSON.stringify(result, null, 2));
    return;
  }

  const taskInfo  = result.task_score ?? {};
  const scoreInfo = result.speech_score ?? {};
  const speechace = scoreInfo.speechace_score ?? {};
  const cefr      = scoreInfo.cefr_score ?? {};
  const ielts     = scoreInfo.ielts_score ?? {};

  const transcript = taskInfo.transcript;
  if (transcript) console.log(`\nTranscript: ${transcript}`);

  const taskScore = taskInfo.score;
  if (taskScore !== undefined && taskScore !== null) {
    console.log(`Task Score: ${taskScore} / 5`);
  }

  const [C1, C2, C3, C4] = [15, 15, 10, 11];
  const SEP = "=".repeat(C1 + C2 + C3 + C4 + 6);

  console.log("\n" + SEP);
  console.log(
    `${"Category".padEnd(C1)}  ${"speechace_score".padStart(C2)}  ${"cefr_score".padStart(C3)}  ${"ielts_score".padStart(C4)}`
  );
  console.log("-".repeat(SEP.length));
  for (const row of ROWS) {
    const s = String(speechace[row] ?? "N/A");
    const c = String(cefr[row] ?? "N/A");
    const i = String(ielts[row] ?? "N/A");
    console.log(
      `${row.padEnd(C1)}  ${s.padStart(C2)}  ${c.padStart(C3)}  ${i.padStart(C4)}`
    );
  }
  console.log(SEP);
}

const { values: args } = parseArgs({
  options: {
    audio:   { type: "string" },
    context: { type: "string" },
    dialect: { type: "string", default: "en-us" },
    raw:     { type: "boolean", default: false },
  },
});

if (!args.audio || !args.context) {
  console.error("Usage: node score_task.js --audio <path> --context <image description>");
  process.exit(1);
}

try {
  const result = await scoreTask(args.audio, args.context, args.dialect);
  if (args.raw) {
    console.log(JSON.stringify(result, null, 2));
  } else {
    printSummary(result);
  }
} catch (err) {
  console.error(err.message);
  process.exit(1);
}

File: basic_samples/python/score_task.py
"""
Speechace API — Task Achievement Scoring
=========================================
Submits an audio file and an image description for task achievement scoring.
Returns a task score (0–5) plus pronunciation, fluency, grammar, vocabulary,
and coherence metrics across multiple scoring frameworks.

Usage:
    python score_task.py --audio ../../audio_samples/apple.wav \
        --context "A photo of a red apple on a white table"

Environment variables (can also be placed in a .env file at the project root):
    SPEECHACE_API_KEY       Your Speechace API key (required)
    SPEECHACE_API_ENDPOINT  Base URL, defaults to https://api.speechace.co
    SPEECHACE_USER_ID       Arbitrary user identifier sent to the API
"""

import argparse
import json
import os
import sys
from pathlib import Path
from urllib.parse import urlencode

import requests
from dotenv import load_dotenv

load_dotenv(Path(__file__).resolve().parents[2] / ".env")

# Score categories returned by the API, in display order
_ROWS = ["overall", "pronunciation", "fluency", "grammar", "vocab", "coherence"]


def build_url(endpoint: str, key: str, dialect: str, user_id: str) -> str:
    base = endpoint.rstrip("/")
    params = urlencode({
        "key": key,
        "dialect": dialect,
        "user_id": user_id,
        "task_type": "describe-image",
    })
    return f"{base}/api/scoring/task/v9/json?{params}"


def score_task(audio_path: str, context: str, dialect: str = "en-us") -> dict:
    """Send an audio file and image description to Speechace for task scoring."""
    key = os.environ.get("SPEECHACE_API_KEY", "")
    endpoint = os.environ.get("SPEECHACE_API_ENDPOINT", "https://api.speechace.co")
    user_id = os.environ.get("SPEECHACE_USER_ID", "sample-user-001")

    if not key:
        raise ValueError(
            "SPEECHACE_API_KEY is not set. "
            "Export it as an environment variable or add it to a .env file."
        )

    url = build_url(endpoint, key, dialect, user_id)

    with open(audio_path, "rb") as audio_file:
        response = requests.post(
            url,
            files={"user_audio_file": audio_file},
            data={"task_context": context, "include_speech_score": "1"},
            timeout=30,
        )

    response.raise_for_status()
    return response.json()


def print_summary(result: dict) -> None:
    """Print a human-readable summary of the task scoring response."""
    status = result.get("status")
    if status != "success":
        print(f"API returned status: {status}")
        print(json.dumps(result, indent=2))
        return

    task_info = result.get("task_score", {})
    transcript = task_info.get("transcript")
    if transcript:
        print(f"\nTranscript: {transcript}")

    task_score = task_info.get("score")
    if task_score is not None:
        print(f"Task Score: {task_score} / 5")

    score_info = result.get("speech_score", {})
    speechace = score_info.get("speechace_score", {})
    cefr      = score_info.get("cefr_score", {})
    ielts     = score_info.get("ielts_score", {})

    C1, C2, C3, C4 = 15, 15, 10, 11
    SEP = "=" * (C1 + C2 + C3 + C4 + 6)

    print(f"\n{SEP}")
    print(f"{'Category':<{C1}}  {'speechace_score':>{C2}}  {'cefr_score':>{C3}}  {'ielts_score':>{C4}}")
    print("-" * len(SEP))
    for row in _ROWS:
        s = speechace.get(row, "N/A")
        c = cefr.get(row, "N/A")
        i = ielts.get(row, "N/A")
        print(f"{row:<{C1}}  {str(s):>{C2}}  {str(c):>{C3}}  {str(i):>{C4}}")
    print(SEP)


def main():
    parser = argparse.ArgumentParser(description="Speechace task achievement scoring sample")
    parser.add_argument("--audio",   required=True, help="Path to audio file (WAV or MP3)")
    parser.add_argument("--context", required=True, help="Image description (task context)")
    parser.add_argument("--dialect", default="en-us", help="Dialect code (default: en-us)")
    parser.add_argument("--raw", action="store_true", help="Print full raw JSON response")
    args = parser.parse_args()

    try:
        result = score_task(args.audio, args.context, args.dialect)
    except ValueError as e:
        print(f"Configuration error: {e}", file=sys.stderr)
        sys.exit(1)
    except requests.HTTPError as e:
        print(f"HTTP error {e.response.status_code}: {e.response.text}", file=sys.stderr)
        sys.exit(1)
    except requests.RequestException as e:
        print(f"Request failed: {e}", file=sys.stderr)
        sys.exit(1)

    if args.raw:
        print(json.dumps(result, indent=2))
    else:
        print_summary(result)


if __name__ == "__main__":
    main()
